Support generating CMS with trusted signing time

emaloney opened the pull request:
Motivation:

We had a need to generate CMS structures with an embedded trusted timestamp that indicates the time at which a signing operation was performed, and discovered that this was not supported by the existing Swift Certificates implementation.

Modifications:

To address this need, we adapted existing functionality, adding several functions allowing the signing time to be embedded in a CMS structure.

In order to make this work, we also needed to add public access to several entities in the code.

Result:

This new CMS SPI allows Swift Certificates to be used for generating trusted timestamp signatures from CMS structures that embed the signing time, such as those for codesigning.

We have been using this exact same code in a high-volume production environment for several months, and can confirm that the CMS structures generated by this are usable for creating valid signatures.

emaloney committed:
rdar://161919434 Upstream `pkg-signer*` support

In order to make `pkg-signer*` work, we needed to make some changes to `swift-certificates` exposing certain entities publicly and adding trusted timestamp support to CMS.

Lukasa commented on Sources/X509/X509BaseTypes/AlgorithmIdentifier.swift:
```diff
@@ -14,27 +14,23 @@
 
-    @usableFromInline
-    var parameters: ASN1Any?
+    public private(set) var parameters: ASN1Any?
```
I don't think there's any particular reason to make these `private(set)`.

emaloney replied:
Changed to just plain `public` in latest version of this PR.

emaloney committed:
Fix `CMSSignerInfo`'s `unsignedAttrs` tag

### Motivation

`CMSSignerInfo` could not round-trip `unsignedAttrs`: it serializes them as
`[1] IMPLICIT SET` but the DER and BER parsers asked for the inner set with
implicit tag `[0]`, so parsing any SignerInfo with unsigned attributes threw
`unexpectedFieldType: [1]`. Fixes issue #306.

### Changes

- Correct the `unsignedAttrs` inner set identifier from `[0]` to `[1]` in the
DER and BER initializers of `CMSSignerInfo`.
- Add a round-trip unit test for a `CMSSignerInfo` carrying `unsignedAttrs`.

### Result

`unsignedAttrs` now round-trip correctly; no behavior change to the (already
correct) serialized output.

emaloney commented on Sources/X509/CryptographicMessageSyntax/CMSEncapsulatedContentInfo.swift:
````diff
@@ -21,10 +21,9 @@ import SwiftASN1
 /// ```
-@usableFromInline
-struct CMSEncapsulatedContentInfo: DERImplicitlyTaggable, BERImplicitlyTaggable, Hashable, Sendable {
+public struct CMSEncapsulatedContentInfo: DERImplicitlyTaggable, BERImplicitlyTaggable, Hashable, Sendable {
````
This change is not present in the latest version of this PR.

emaloney commented on Sources/X509/CryptographicMessageSyntax/CMSVersion.swift:
```diff
@@ -19,27 +19,25 @@ import SwiftASN1
-struct CMSVersion: RawRepresentable, Hashable, Sendable {
-    @usableFromInline
-    var rawValue: Int
+public struct CMSVersion: RawRepresentable, Hashable, Sendable {
```
This change is not present in the latest version of this PR.

emaloney commented on Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:
```diff
@@ -78,39 +78,106 @@ public enum CMS: Sendable {
+
+    @_spi(CMS)
+    @inlinable
+    public static func createSigningTimeASN1(signingTime: Date) throws -> Data {
```
This change is not present in the latest version of this PR.

emaloney commented on Sources/X509/Digests.swift:
```diff
@@ -19,16 +19,15 @@ import Foundation
-@usableFromInline
 @available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, macCatalyst 13, visionOS 1.0, *)
-enum Digest: Sendable {
+public enum Digest: Sendable {
```
I see that since my original PR was opened, `Digest` has gone from an `enum` to a `struct`, and the cases you were (presumably) worried about exposing are now hidden in the inner `enum` type `Backing`.

Am I correct in assuming that making `Digest` and the `computeDigest(for:using:)` function `public` is no longer a concern given this refactoring?

Lukasa commented on Sources/X509/CryptographicMessageSyntax/CMSOperations.swift:
```diff
@@ -80,39 +80,96 @@ public enum CMS: Sendable {
 
+    @_spi(CMS)
+    @inlinable
+    public static func getSignedAttributesBytes<Data: DataProtocol>(
```
Is there a reason this needs to be an SPI?

emaloney replied:
No technical reason; I will remove that.

emaloney replied:
Pushed an update that removes the `@_spi(CMS)` — note that the `signWithTrustedTimestamp` function is still marked as SPI. I left that as-is b/c I wasn't sure if that signature was something we were ready to formalize & make available publicly.
